snippet_search/snippet_search.py

```diff
@@ -0,0 +1,151 @@
+import re
+from collections import Counter, defaultdict
+from typing import Optional
+
+import discord
+from discord.ext import commands
+
+from bot import ModmailBot
+from core import checks
+from core.models import PermissionLevel
+from core.paginator import EmbedPaginatorSession
+from core.utils import escape_code_block, truncate
+
+
+WORD_PATTERN = re.compile(r"[a-zA-Z]+")
+THRESHOLD = 1.0
+
+
+def score(query: Optional[str], name: str, content: str) -> float:
+    """
+    Return a numerical sorting score for a snippet based on a query.
+
+    More relevant snippets have higher scores. If the query is None,
+    return a score that always meets the search inclusion threshold.
+    """
+    if query is None:
+        return THRESHOLD
+    return (
+        (common_word_count(query, name) + common_word_count(query, content))
+        / len(words(query))
+    )
+
+
+def words(s: str) -> list[str]:
+    """
+    Extract a list of 'words' from the given string.
+
+    A 'word' is defined by the WORD_PATTERN regex.  This is purely for
+    use by the scoring function so isn't perfect.
+    """
+    return WORD_PATTERN.findall(s)
+
+
+def common_word_count(s1: str, s2: str) -> int:
+    """Return the number of words in common between the two strings."""
+    return sum(
+        (
+            Counter(map(str.casefold, words(s1)))
+            & Counter(map(str.casefold, words(s2)))
+        ).values()
+    )
+
+
+def group_snippets_by_content(snippets: dict[str, str]) -> list[tuple[set[str], str]]:
+    """
+    Take a dictionary of snippets (in the form {name: content}) and group together snippets with the same content.
+
+    Snippet contents are stipped of leading and trailing whitespace
+    before comparison.
+
+    The result is of the form [(set_of_snippet_names, content)].
+    """
+    names_by_content = defaultdict(set)
+    for name, content in snippets.items():
+        names_by_content[content.strip()].add(name)
+    grouped_snippets = []
+    for group in names_by_content.values():
+        name, *_ = group
+        content = snippets[name]
+        grouped_snippets.append((group, content))
+    return grouped_snippets
+
+
+class SnippetSearch(commands.Cog):
+    """A plugin that provides a command for searching snippets."""
+
+    def __init__(self, bot: ModmailBot):
+        self.bot = bot
+
+    @checks.has_permissions(PermissionLevel.SUPPORTER)
+    @commands.command(name="snippetsearch")
+    async def snippet_search(
+        self, ctx: commands.Context, *, query: Optional[str] = None
+    ) -> None:
+        """Search for a snippet."""
+        grouped_snippets = group_snippets_by_content(self.bot.snippets)
+
+        scored_groups = []
+        for i, (names, content) in enumerate(grouped_snippets):
+            group_score = max(score(query, name, content) for name in names)
+            scored_groups.append((group_score, i, names, content))
+
```

Comment:
Looks like the enumerate is to ensure tuple sorting is done properly in the case of equal scores, idk how much it matters particularly tho

Could replace 89-99 with:
```py
        for i, (names, content) in enumerate(grouped_snippets):
            group_score = max(score(query, name, content) for name in names)
            #or score(query, names, content) if modifying score

            if group_score >= THRESHOLD: #saves sorting time?
                scored_groups.append((group_score, i, names, content))

        scored_groups.sort(reverse=True)
```
Line 116 `for _, _, names, content in matching_snippet_groups` would avoid the need for lines 95-99.